### PR TITLE
Fixed #31894 -- Added note and tests for JSONField key lookups with QuerySet.exclude().

### PR DESCRIPTION
**Pull request by jpribyl**
[ticket-31894](https://code.djangoproject.com/ticket/31894)

Thanks for reporting this @carltongibson and @laymonage !

### CODE REVIEW

File: docs/topics/db/queries.txt [hunk 1]
```diff
@@ -937,6 +937,42 @@ To query for missing keys, use the ``isnull`` lookup::
...
+    >>> Dog.objects.count()
+    4
+    >>> Dog.objects.filter(data__breed='collie').count()
+    2
+    >>> Dog.objects.exclude(data__breed='collie').count()
+    1
+
+A negated filter will still return the same objects as an exclude::
+
+    >>> Dog.objects.filter(~Q(data__owner='Bob'))
```
Comment (laymonage) [inline docs/topics/db/queries.txt:971]: Should we also note that the query that returns a proper negation of `data__owner='Bob'` would be `~Q(data__owner='Bob') | Q(data__owner__isnull=True)`?

File: tests/queries/tests.py [hunk 2]
```diff
@@ -2878,6 +2878,54 @@ def test_exclude_nullable_fields(self):
         )
 
 
+@skipUnlessDBFeature('supports_json_field')
+class ExcludeJSONFieldTests31894(TestCase):
```
Comment (laymonage) [inline tests/queries/tests.py:2882]: I'm not sure whether this should be here or in [tests/model_fields/test_jsonfield.py](https://github.com/django/django/blob/master/tests/model_fields/test_jsonfield.py).

**Review by laymonage — CHANGES_REQUESTED** [review]
Thanks for the PR! I left some comments.

File: docs/topics/db/queries.txt (hunk 1 above)
Comment (jpribyl) [reply docs/topics/db/queries.txt:971]: Absolutely! Fixed

File: tests/queries/tests.py (hunk 2 above)
Comment (jpribyl) [reply tests/queries/tests.py:2882]: I'm not sure either :thinking: Probably it doesn't matter too much? Happy to move them if it seems important though!

File: docs/topics/db/queries.txt [hunk 3]
```diff
@@ -937,6 +937,42 @@ To query for missing keys, use the ``isnull`` lookup::
     On PostgreSQL, if only one key or index is used, the SQL operator ``->`` is
     used. If multiple operators are used then the ``#>`` operator is used.
 
+Querying optional JSON keys
+--------------------------------------
```
[the remark this answers is not in the document]
Comment (jpribyl) [reply docs/topics/db/queries.txt]: Fixed!

File: tests/queries/tests.py (hunk 2 above)
Comment (felixxm) [reply tests/queries/tests.py:2882]: Please move these tests to `tests.model_fields.test_jsonfield.TestQuerying`. We don't need a separate class. Also, reuse existing objects is tests.

File: tests/queries/tests.py [hunk 4]
```diff
@@ -2878,6 +2878,54 @@ def test_exclude_nullable_fields(self):
...
+    def test_exclude_json_field_does_not_return_null_keys(self):
+        q = Q(json_field__name='bigfoot')
+        self.assertQuerysetEqual(
+            list(JSONFieldNullable.objects.exclude(q)),
+            ['<JSONFieldNullable: JSONFieldNullable object (2)>']
+        )
+        self.assertQuerysetEqual(
+            list(JSONFieldNullable.objects.exclude(~q)),
+            ['<JSONFieldNullable: JSONFieldNullable object (1)>']
+        )
```
Comment (felixxm) [inline tests/queries/tests.py:2926]: For example:
```suggestion
        self.assertSequenceEqual(
            JSONFieldNullable.objects.exclude(json_field__name='bigfoot'),
            [self.some_obj_1],
        )
        self.assertSequenceEqual(
            JSONFieldNullable.objects.exclude(~Q(json_field__name='bigfoot')),
            [self.some_obj_2],
        )
```

File: docs/topics/db/queries.txt [hunk 5]
```diff
@@ -937,6 +937,48 @@ To query for missing keys, use the ``isnull`` lookup::
...
+    <QuerySet [<Dog: Max>]>
+    >>> Dog.objects.exclude(data__owner='Bob')
+    <QuerySet [<Dog: Max>]>
+
+It is possible to return all objects which match the negated filter or have a missing
+key like this:
+
+    >>> Dog.objects.filter(~Q(data__owner='Bob') | Q(data__owner__isnull=True))
+    <QuerySet [<Dog: Meg>, <Dog: Archie>, <Dog: Max>]>
+
```
Comment (felixxm) [inline docs/topics/db/queries.txt:981]: IMO we don't need a separate section a small warning should be enough, e.g.
```
.. warning::

    Due to the way in which key-path queries work, ``exclude()`` and ``filter``
    are not guaranteed to produce exhaustive sets.

**Review by felixxm — COMMENTED** [review]
@jpribyl Thanks for this patch :+1:

**Comment by jpribyl** [comment]
@felixxm Thanks for the suggestions! I believe I have fixed those issues, do you see anything else or is this one good to go?

**Comment by felixxm** [comment]
@jpribyl Thanks :+1: Welcome aboard :boat: 

I simplified tests and pushed minor edits to docs.

**Comment by laymonage** [comment]
Should we also add tests that demonstrate adding the `isnull` lookup would result in an exhaustive set?

**Comment by felixxm** [comment]
@laymonage Added.

**Pull request merged**